hbase-server/src/test/java/org/apache/hadoop/hbase/io/hfile/bucket/TestBucketCache.java

```diff
@@ -255,51 +269,128 @@ public void run() {
     BucketCache bucketCache = new BucketCache(ioEngineName, capacitySize, constructedBlockSize,
-        constructedBlockSizes, writeThreads, writerQLen, persistencePath);
-    long usedSize = bucketCache.getAllocator().getUsedSize();
-    assertEquals(0, usedSize);
-
-    HFileBlockPair[] blocks = CacheTestUtils.generateHFileBlocks(constructedBlockSize, 1);
-    // Add blocks
-    for (HFileBlockPair block : blocks) {
-      bucketCache.cacheBlock(block.getBlockName(), block.getBlock());
-    }
-    for (HFileBlockPair block : blocks) {
-      cacheAndWaitUntilFlushedToBucket(bucketCache, block.getBlockName(), block.getBlock());
+            smallBucketSizes, writeThreads, writerQLen, persistencePath);
+    assertFalse(new File(persistencePath).exists());
+    assertEquals(0, bucketCache.getAllocator().getUsedSize());
+    assertEquals(0, bucketCache.backingMap.size());
+    HBASE_TESTING_UTILITY.cleanupTestDir();
+  }
+
+  private void testRetrievalUtils(Path testDir, String ioEngineName)
+          throws IOException, InterruptedException {
+    final String persistencePath = testDir + "/bucket.persistence";
+    BucketCache bucketCache = new BucketCache(ioEngineName, capacitySize, constructedBlockSize,
+            constructedBlockSizes, writeThreads, writerQLen, persistencePath);
+    try {
+      long usedSize = bucketCache.getAllocator().getUsedSize();
+      assertEquals(0, usedSize);
+      HFileBlockPair[] blocks = CacheTestUtils.generateHFileBlocks(constructedBlockSize, 1);
+      for (HFileBlockPair block : blocks) {
+        bucketCache.cacheBlock(block.getBlockName(), block.getBlock());
+      }
+      for (HFileBlockPair block : blocks) {
+        cacheAndWaitUntilFlushedToBucket(bucketCache, block.getBlockName(), block.getBlock());
+      }
+      usedSize = bucketCache.getAllocator().getUsedSize();
+      assertNotEquals(0, usedSize);
+      bucketCache.shutdown();
+      assertTrue(new File(persistencePath).exists());
+      bucketCache = new BucketCache(ioEngineName, capacitySize, constructedBlockSize,
+              constructedBlockSizes, writeThreads, writerQLen, persistencePath);
+      assertFalse(new File(persistencePath).exists());
+      assertEquals(usedSize, bucketCache.getAllocator().getUsedSize());
+    } finally {
+      bucketCache.shutdown();
     }
-    usedSize = bucketCache.getAllocator().getUsedSize();
-    assertNotEquals(0, usedSize);
-    // persist cache to file
-    bucketCache.shutdown();
     assertTrue(new File(persistencePath).exists());
+  }
 
-    // restore cache from file
-    bucketCache = new BucketCache(ioEngineName, capacitySize, constructedBlockSize,
-        constructedBlockSizes, writeThreads, writerQLen, persistencePath);
-    assertFalse(new File(persistencePath).exists());
-    assertEquals(usedSize, bucketCache.getAllocator().getUsedSize());
-    // persist cache to file
-    bucketCache.shutdown();
-    assertTrue(new File(persistencePath).exists());
+  @Test
```

Comment:
@Test(expected = IllegalArgumentException.class) ?

Reply:
Apart from this, the patch lgtm.

Reply:
It is removed so that we can specifically assert the IAE error message here: https://github.com/apache/hbase/pull/333/files#diff-ab0037c891480f9136f0e222ba490035R344

Reply:
Perfect! +1

Reply:
Thanks a lot for your review @jatsakthi